minio-admin/src/main/java/io/minio/admin/MinioAdminClient.java

```diff
@@ -0,0 +1,523 @@
+   * @throws NoSuchAlgorithmException thrown to indicate missing of MD5 or SHA-256 digest library.
+   * @throws InvalidKeyException thrown to indicate missing of HMAC SHA-256 library.
+   * @throws IOException thrown to indicate I/O error on S3 operation.
+   */
+  public void removeCannedPolicy(String policyName)
+      throws InternalException, InsufficientDataException, NoSuchAlgorithmException,
+          InvalidKeyException, IOException {
+    executeAdmin(
+        Method.DELETE,
+        "remove-canned-policy",
+        region,
+        null,
+        ImmutableMultimap.of("name", policyName),
+        null,
+        0);
+  }
+
+  /** Build URL for given parameters. */
+  protected HttpUrl buildAdminUrl(String action, Multimap<String, String> queryParamMap) {
+
+    HttpUrl.Builder urlBuilder = this.baseUrl.newBuilder();
+    String host = this.baseUrl.host();
+
+    urlBuilder.host(host);
+    urlBuilder.addEncodedPathSegment(S3Escaper.encode("minio"));
+    urlBuilder.addEncodedPathSegment(S3Escaper.encode("admin"));
+    urlBuilder.addEncodedPathSegment(S3Escaper.encode("v3"));
+    urlBuilder.addEncodedPathSegment(S3Escaper.encode(action));
+
+    if (queryParamMap != null) {
+      for (Map.Entry<String, String> entry : queryParamMap.entries()) {
+        urlBuilder.addEncodedQueryParameter(
+            S3Escaper.encode(entry.getKey()), S3Escaper.encode(entry.getValue()));
+      }
+    }
+
+    return urlBuilder.build();
+  }
+
+  /** Execute HTTP request for given parameters. */
+  protected Response executeAdmin(
+      Method method,
+      String action,
+      String region,
+      Headers headers,
+      Multimap<String, String> queryParamMap,
+      Object body,
+      int length)
+      throws InsufficientDataException, InternalException, InvalidKeyException, IOException,
+          NoSuchAlgorithmException {
+
+    if (body != null && !(body instanceof byte[])) {
+      body = OBJECT_MAPPER.writeValueAsString(body);
+    }
+
+    if (body == null && (method == Method.PUT || method == Method.POST)) body = EMPTY_BODY;
+
+    HttpUrl url = buildAdminUrl(action, queryParamMap);
+    Credentials creds = (provider == null) ? null : provider.fetch();
+    Request request = createRequest(url, method, headers, body, length, creds);
+    if (creds != null) {
+      request =
+          Signer.signV4S3(
+              request,
+              region,
+              creds.accessKey(),
+              creds.secretKey(),
+              request.header("x-amz-content-sha256"));
+    }
+
+    OkHttpClient httpClient = this.httpClient;
+
+    Response response = httpClient.newCall(request).execute();
+
+    if (response.code() != 200) {
+      throw new RuntimeException("Request failed with response: " + response.body().string());
+    }
+
+    return response;
+  }
+
+  public static MinioAdminClient.Builder builder() {
+    return new MinioAdminClient.Builder();
+  }
+
+  public static final class Builder {
```

Comment:
A lot of this is copied from MinioClient -- I'm happy to throw some of the validation methods into a shared utility class? Could also maybe create a parent class called (MinioClientBuilder) or something as well.

Reply:
1. As this class is `MinIO` specific, you can trim lot of code here e.g. we don't need to check whether given endpoint is AWS S3 etc.
2. Feel free to make `protected` class/methods in `MinioClient` and use it here.

Reply:
What I ended up doing here, which greatly reduced the amount of code, is just made a `BaseBuilder<T>` in the S3Base.

One hard issue with using the protected methods as you suggested, is that the class is `static`, so it does not have a reference to the current object (as it does not exist). I guess the other alternative is to make the functions `protected static`?